File: mcp/llm_provider/anthropic.rb
require 'anthropic'
require_relative 'base'

module MCP
  module LLMProvider
    class Anthropic < Base
      def initialize(options = {})
        super
        api_key = options[:api_key] || ENV.fetch('ANTHROPIC_API_KEY', nil)
        @client = ::Anthropic::Client.new(
          access_token: api_key,
          log_errors: true
        )
        @model = options[:model] || 'claude-3-7-sonnet-20250219'
        @system = options[:system] || 'Respond only in Japanese.'
      end

      def generate_message(messages:, tools:, max_tokens: 1000)
        @client.messages(
          parameters: {
            model: @model,
            system: @system,
            messages: messages,
            max_tokens: max_tokens,
            tools: tools
          }
        )
      end

      def extract_text(response)
        return '' if response['content'].empty?

        text_content = response['content'].select { |content| content['type'] == 'text' }
        text_content.map { |content| content['text'] }.join("\n")
      end

      def extract_tool_use(response)
        return nil if response['content'].empty?

        tool_use = response['content'].find { |content| content['type'] == 'tool_use' }
        return nil unless tool_use

        {
          id: tool_use['id'],
          name: tool_use['name'],
          args: tool_use['input']
        }
      end

      def create_tool_result_message(messages:, tool_use:, result:)
        # まず最新のassistantメッセージを抽出
        assistant_message_content = []
        response_content = messages.last[:content] if messages.last[:role] == 'assistant'
        
        if response_content
          # 既存のassistantメッセージを更新
          assistant_message_content = response_content
        else
          # 新規assistantメッセージを作成してメッセージ配列に追加
          assistant_message_content = [{ 'type' => 'tool_use', 'id' => tool_use[:id], 'name' => tool_use[:name], 'input' => tool_use[:args] }]
          messages << { 'role' => 'assistant', 'content' => assistant_message_content }
        end

        # ツール結果メッセージを追加
        messages << {
          'role' => 'user',
          'content' => [
            {
              'type' => 'tool_result',
              'tool_use_id' => tool_use[:id],
              'content' => result.to_s
            }
          ]
        }

        messages
      end
    end
  end
end 

File: wip-typescript/src/llm-provider/anthropic.ts
import Anthropic from '@anthropic-ai/sdk';
import { BaseLLMProvider } from './base-provider.js';
import { LLMMessage, LLMOptions } from '../types/llm.js';

/**
 * Anthropic LLMプロバイダー
 */
export class AnthropicProvider extends BaseLLMProvider {
  readonly name = 'anthropic';
  private client: Anthropic;
  private defaultModel = 'claude-3-sonnet-20240229';

  /**
   * コンストラクタ
   * 環境変数からAPIキーを取得
   * @throws APIキーが設定されていない場合はエラー
   */
  constructor() {
    super();
    const apiKey = process.env.ANTHROPIC_API_KEY;
    if (!apiKey) {
      throw new Error('Anthropic API key is required');
    }
    this.client = new Anthropic({ apiKey });
  }

  /**
   * Anthropic APIにメッセージを送信して応答を取得
   * @param messages 送信するメッセージの配列
   * @param options LLM呼び出しオプション
   * @returns Anthropicからの応答テキスト
   */
  async sendMessage(messages: LLMMessage[], options: LLMOptions = {}): Promise<string> {
    try {
      const temperature = options.temperature ?? 0.7;
      const maxTokens = options.maxTokens ?? 4096;
      
      // システムメッセージを抽出（最初のsystemメッセージを使用）
      const systemMessage = messages.find(m => m.role === 'system')?.content || '';
      
      // システムメッセージ以外のメッセージを抽出
      const nonSystemMessages = messages.filter(m => m.role !== 'system');
      
      // Anthropicの型に変換
      const anthropicMessages = nonSystemMessages.map(msg => {
        if (msg.role === 'function') {
          // Anthropicはfunction messagesをサポートしていないためassistantとして扱う
          return { role: 'assistant' as const, content: msg.content };
        }
        
        // roleの型を明示的に指定
        let role: 'user' | 'assistant';
        if (msg.role === 'user') {
          role = 'user';
        } else if (msg.role === 'assistant') {
          role = 'assistant';
        } else {
          // デフォルトはuserとして扱う
          role = 'user';
        }
        
        return { role, content: msg.content };
      });
      
      const response = await this.client.messages.create({
        model: this.defaultModel,
        ...(systemMessage ? { system: systemMessage } : {}),
        messages: anthropicMessages,
        temperature,
        max_tokens: maxTokens
      });

      // 応答テキストを取得
      const textContent = response.content
        .filter(content => content.type === 'text')
        .map(content => 'text' in content ? content.text : '')
        .join('\n');

      return textContent;
    } catch (error) {
      throw new Error(`Anthropic API error: ${error instanceof Error ? error.message : String(error)}`);
    }
  }
} 